core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.kafka.internal
+
+import akka.NotUsed
+import akka.kafka.Subscription.{TopicSubscription, TopicSubscriptionPattern}
+import akka.kafka.scaladsl.Consumer.Control
+import akka.kafka.scaladsl.KafkaAsyncConsumer
+import akka.kafka.{AutoSubscription, ConsumerSettings}
+import akka.stream.scaladsl.Source
+import akka.stream.stage._
+import akka.stream.{ActorMaterializer, Attributes, Outlet, SourceShape}
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.util.{Failure, Success, Try}
+
+private[kafka] abstract class SubSourceLogic[K, V, Msg](
+    shape: SourceShape[(TopicPartition, Source[Msg, NotUsed])],
+    settings: ConsumerSettings[K, V],
+    subscription: AutoSubscription
+) extends GraphStageLogic(shape) with PromiseControl with MessageBuilder[K, V, Msg] {
+  var consumer: KafkaAsyncConsumer[K, V] = _
+  var buffer: immutable.Queue[TopicPartition] = immutable.Queue.empty
+  var subSources: Map[TopicPartition, Control] = immutable.Map.empty
+  implicit lazy val ec = ActorMaterializer.downcast(materializer).executionContext
+
+  override def preStart(): Unit = {
+    super.preStart()
+    consumer = KafkaAsyncConsumer[K, V](settings)(ActorMaterializer.downcast(materializer).system)
+    subscription match {
+      case TopicSubscription(topics) =>
+        consumer.subscribe(topics, KafkaAsyncConsumer.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+      case TopicSubscriptionPattern(topics) =>
+        consumer.subscribe(topics, KafkaAsyncConsumer.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+    }
+  }
+
+  val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
+    buffer = buffer.enqueue(tps.toList)
+    pump()
+  }
+  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { r =>
+    r.map(subSources.get).foreach(_.foreach(_.shutdown()))
+    subSources --= r
+  }
+
+  val subsourceCancelledCB = getAsyncCallback[TopicPartition] { tp =>
+    subSources -= tp
+    buffer :+= tp
```

Comment:
Looks strange, but we really have no way to control list of assigned partitions. We are responsible for partitions which kafka server assigns to the client. I decide that in case of cancellation I emit partition again.